### PR TITLE
fix ping on OSX when running as root

### LINKED ISSUE #33430: SendPingAsync: Exception when using as root under macOS
Hello,

when using `SendPingAsync` as root on a macOS system, a SocketException is thrown (`System.Net.Sockets.SocketException (56): Socket is already connected`). If run as a regular user, it works. It also works for root if you use .NET 2.2. This leads me to believe that this behaviour was introduced in this commit c6efaf4ef89902f0f2180dba9d5d7be9065d80bb. I also tested this on different macOS machines and the outcome was the same, but on a Ubuntu machine it works with sudo as well.

Here is the program I used to reproduce this issue: https://gist.github.com/pgrosslicht/0acf36322247793925dd62692b9202cd.
Here are screenshots running said program with .NET 3.1 with and without sudo and with .NET 2.2 with and without sudo.

NET3 1
NET2 2

Thanks in advance!

### LINKED ISSUE #34879: icmp ping broken on osx under su account with PingOptions
icmp ping broken on osx under su account

# General

in a dotnet core app calling ping works if not running under the super user account. If you run it as root it fails with operation not supported errors.

using dotnet core v3.1.201 on OSX Mojave

sample app provided. Output:

seans-Mac-mini:Agent sean$ dotnet testping.dll
found 192.168.15.1
found 192.168.15.7
found 192.168.15.0
found 192.168.15.17
found 192.168.15.10
found 192.168.15.19
found 192.168.15.9
found 192.168.15.24
found 192.168.15.12
found 192.168.15.20
found 192.168.15.6
found 192.168.15.5
found 192.168.15.18
Found 13 addresses

seans-Mac-mini:Agent sean$ **sudo** dotnet testping.dll

ping: 192.168.15.3 :System.AggregateException: One or more errors occurred. (An exception occurred during a Ping request.)
 ---> System.Net.NetworkInformation.PingException: An exception occurred during a Ping request.
 ---> System.Net.Sockets.SocketException (45): Operation not supported
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue, Boolean silent)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue)
   at System.Net.Sockets.Socket.set_DontFragment(Boolean value)
   at System.Net.NetworkInformation.Ping.GetRawSocket(SocketConfig socketConfig)
   at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   --- End of inner exception stack trace ---
...
[testping.zip](https://github.com/dotnet/core/files/4467432/testping.zip)

### PR DESCRIPTION
**Pull request by wfurt**
This is fix for thee distance issues with current implementation.

first is regression on OSX caused by https://github.com/dotnet/corefx/pull/34084
It seems like OSX does connect behind the scenes and Send() files `System.Net.Sockets.SocketException (56): Socket is already connected` even if we call `Connect` just once. So I limit that to Linux only. 
fixes #33430

Second issues is when PingOptions is used. In that case we would always try to set ` socket.DontFragment = socketConfig.Options.DontFragment` and that fails on OSX because that is not supported by OS. Since we use RAW socket we have option too include IP header and set DF bit directly. I used Wireshark to verify that DF is properly set based on PingOptions.
fixes #34879

Third problem is with IPv6: (currently not tracked by any issue)
```

          Child exception:
            System.Net.NetworkInformation.PingException: An exception occurred during a Ping request.
           ---> System.PlatformNotSupportedException: This platform does not support packet information for dual-mode sockets.  If packet information is not required, use Socket.Receive.  If packet information is required set Socket.DualMode to false.
          /Users/furt/github/wfurt-runtime/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs(40,0): at System.Net.Sockets.SocketPal.CheckDualModeReceiveSupport(Socket socket)
          /Users/furt/github/wfurt-runtime/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(1744,0): at System.Net.Sockets.Socket.ReceiveFrom(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, EndPoint& remoteEP)
          /Users/furt/github/wfurt-runtime/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(1812,0): at System.Net.Sockets.Socket.ReceiveFrom(Byte[] buffer, SocketFlags socketFlags, EndPoint& remoteEP)
          /Users/furt/github/wfurt-runtime/src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs(206,0): at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocket(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
```

Since we know what AF we need, the fix was to turn off DualMode on RAW socket. 
I don't think we really need "packet information" and that can be possibly future improvement. 

All this happened because we do not have any test coverage for RAW sockets in practice since CI runs as ordinary user. However, some other assemblies use RemoteExcetopr with RunAsSudo to get privileged access. It seems like we should be able to detect in CI. My prototype runs shows exactly the reported issues so it may be worth of getting at least some basic coverage. 
https://helix.dot.net/api/2019-06-17/jobs/353faf88-9320-4dfa-8215-534c515b3fa0/workitems/System.Net.Ping.Functional.Tests/console

Downside of this is that the test depends now on sudo for outer loop runs. It may asked for password if executed interactively or it may fail if given user is not in shudders group.
Given the history, it seems like this may be worth oof some hassle.

### CODE REVIEW

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 1]
```diff
@@ -19,6 +19,8 @@ public partial class Ping
         private const int IcmpHeaderLengthInBytes = 8;
         private const int MinIpHeaderLengthInBytes = 20;
         private const int MaxIpHeaderLengthInBytes = 60;
+        private static bool _sendIpHeader = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
```
Comment (gfoidl) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:22]:
```suggestion
        private static readonly bool _sendIpHeader = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
```
Same below for `_needsConnect`.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 2]
```diff
@@ -19,6 +19,8 @@ public partial class Ping
...
+        private static bool _needsConnect = RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:23]: These are very fast calls; I'm not sure they need to be cached like this.  But if you think it's important, make 'em readonly as suggested.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 3]
```diff
@@ -68,26 +87,38 @@ private SocketConfig GetSocketConfig(IPAddress address, byte[] buffer, int timeo
...
-
-            // Setting Socket.DontFragment and .Ttl is not supported on Unix, so socketConfig.Options is ignored.
             AddressFamily addrFamily = ep.Address.AddressFamily;
+
             Socket socket = new Socket(addrFamily, SocketType.Raw, socketConfig.ProtocolType);
             socket.ReceiveTimeout = socketConfig.Timeout;
             socket.SendTimeout = socketConfig.Timeout;
+            if (addrFamily == AddressFamily.InterNetworkV6)
+            {
+                socket.DualMode = false;
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:97]: Why is this necessary?  Isn't it only set to true by the two-arg Socket ctor that's not being used here?

Comment (stephentoub) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:97]: Ah, nevermind, I guess it's necessary because it's actually calling setsockopt.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 4]
```diff
@@ -395,21 +444,34 @@ public SocketConfig(EndPoint endPoint, int timeout, PingOptions? options, bool i
...
+        private static unsafe byte[] CreateSendMessageBuffer(IpHeader ipHeader, IcmpHeader icmpHeader, byte[] payload)
         {
-            int headerSize = sizeof(IcmpHeader);
-            byte[] result = new byte[headerSize + payload.Length];
-            Marshal.Copy(new IntPtr(&header), result, 0, headerSize);
-            payload.CopyTo(result, headerSize);
-            ushort checksum = ComputeBufferChecksum(result);
+            int icmpHeaderSize = sizeof(IcmpHeader);
+            int offset = 0;
+            int packetSize = ipHeader.TotalLength != 0 ? ipHeader.TotalLength : icmpHeaderSize + payload.Length;
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:451]:
```suggestion
            int packetSize = ipHeader.TotalLength != 0 ? ipHeader.TotalLength : checked(icmpHeaderSize + payload.Length);
```
?  That way if it does overflow at least a more useful exception will be thrown.  Not a big deal, though.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 5]
```diff
@@ -395,21 +444,34 @@ public SocketConfig(EndPoint endPoint, int timeout, PingOptions? options, bool i
...
+            byte[] result = new byte[packetSize];
+
+            if (ipHeader.TotalLength != 0)
+            {
+                int ipHeaderSize = sizeof(IpHeader);
+                Marshal.Copy(new IntPtr(&ipHeader), result, 0, ipHeaderSize);
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:457]: FWIW, you could also do this with span:
```C#
new Span<byte>(&ipHeader, sizeof(IpHeader)).CopyTo(result);
```

File: src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs [hunk 6]
```diff
@@ -880,5 +886,37 @@ public async Task Ping_TimedOut_TAP_Success()
...
+            {
+                byte[] buffer = TestSettings.PayloadAsBytes;
+                SendBatchPing(
+                    (ping) => ping.Send(address, TestSettings.PingTimeout, buffer, new PingOptions()),
+                    (pingReply) =>
+                    {
+                        PingResultValidator(pingReply, new IPAddress[] { IPAddress.Parse(address) }, null);
+                        Assert.Equal(buffer, pingReply.Buffer);
+                    });
+            }, localIpAddress.ToString(), new RemoteInvokeOptions { RunAsSudo = true }))
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs:917]: You can just use `.Dispose()` rather than the empty using block.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 7]
```diff
@@ -46,19 +48,36 @@ private async Task<PingReply> SendPingAsyncCore(IPAddress address, byte[] buffer
...
+            IpHeader iph = default;
 
             bool ipv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            bool sendIpHeader = ipv4 && options != null && _sendIpHeader;
+
+             if (sendIpHeader)
+             {
+                iph.VersionAndLength = 0x45;
+                // On OSX this strangely must be host byte order.
+                iph.TotalLength = (ushort)(sizeof(IpHeader) + sizeof(IcmpHeader) +  buffer.Length);
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs:66]:
```suggestion
                iph.TotalLength = checked((ushort)(sizeof(IpHeader) + sizeof(IcmpHeader) +  buffer.Length));
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs (hunk 6 above)
Comment (wfurt) [reply src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs]: I like that better. I copy test structure from other assembly.
